File: maliki/maliki/overrides/stock_entry.py
import frappe
from frappe.model.mapper import get_mapped_doc
from erpnext.stock.doctype.stock_entry.stock_entry import (StockEntry)
from frappe.utils import flt

class CustomStockEntry(StockEntry):
    def on_submit(self):
        self.register_additional_costs_je()
        super().on_submit()
    
    def on_cancel(self):
        super().on_cancel()
        if self.additional_costs_je:
            doc = frappe.get_doc("Journal Entry", self.additional_costs_je)
            doc.cancel()

    def before_submit(self):
        tr_warehouse = frappe.db.get_single_value("Maliki Settings", "tr_warehouse")

        if self.to_warehouse == tr_warehouse and self.total_additional_costs == 0:
            frappe.throw('يجب تحديد تكلفة النقل')

    @frappe.whitelist()
    def set_custom_status(self):
        total_qty = sum([d.qty for d in self.items])
        parent_doc = frappe.get_doc("Stock Entry", self.outgoing_stock_entry)
        issues_qty = total_qty + parent_doc.issues_qty

        parent_doc.db_set('issues_qty', issues_qty, commit=True)
        parent_doc.db_set('transfer_status', self.transfer_status, commit=True)

        parent_doc.add_comment('Comment', f'تم نحويل عدد {total_qty} من المنتجات لمخزن الأضرار بسبب {self.issue_type}')
        frappe.db.commit()
    
    @frappe.whitelist()
    def register_additional_costs_je(self):
        if self.total_additional_costs > 0:
            settings_doc = frappe.get_doc("Maliki Settings")
            shipping_account = settings_doc.shipping_account
            tr_account = settings_doc.tr_account

            if not shipping_account:
                frappe.throw("Shipping Costs Account must be set in Maliki Settings")
            if not tr_account:
                frappe.throw("Turkey Account must be set in Maliki Settings")
            
            doc = frappe.new_doc('Journal Entry')
            doc.title = f"مصاريف شحن {self.name}"

            doc.voucher_type = "Journal Entry"
            # doc.user_remark = self.notes
            doc.posting_date = self.posting_date

            # money to account
            to_ = {"account":shipping_account,
                    # "cost_center": self.cost_center,
                    "debit_in_account_currency": self.total_additional_costs}
            # money from account
            from_ = {"account":tr_account,
                    # "cost_center": self.cost_center,
                    "credit_in_account_currency": self.total_additional_costs}
            
            doc.append("accounts",to_)
            doc.append("accounts",from_)

            doc.save(ignore_permissions=True)
            doc.submit()

            self.db_set('additional_costs_je', doc.name)
            frappe.msgprint(f"Journal Entry {doc.name} created")
        else:
            frappe.msgprint("لا يوجد تكلفة شحن")


@frappe.whitelist()
def make_dmg_stock_in_entry(source_name, target_doc=None):
    to_warehouse = frappe.db.get_single_value("Maliki Settings", "damaged_warehouse")
    def set_missing_values(source, target):
        target.stock_entry_type = "Material Transfer"
        target.set_missing_values()

        if not frappe.db.get_single_value("Stock Settings", "use_serial_batch_fields"):
            target.make_serial_and_batch_bundle_for_transfer()

    def update_item(source_doc, target_doc, source_parent):
        target_doc.t_warehouse = to_warehouse

        if source_doc.material_request_item and source_doc.material_request:
            add_to_transit = frappe.db.get_value("Stock Entry", source_name, "add_to_transit")
            if add_to_transit:
                warehouse = frappe.get_value(
                    "Material Request Item", source_doc.material_request_item, "warehouse"
                )
                target_doc.t_warehouse = warehouse

        target_doc.s_warehouse = source_doc.t_warehouse
        target_doc.qty = source_doc.qty - source_doc.transferred_qty

    doclist = get_mapped_doc(
        "Stock Entry",
        source_name,
        {
            "Stock Entry": {
                "doctype": "Stock Entry",
                "field_map": {"name": "outgoing_stock_entry", 
                  "to_warehouse": "from_warehouse",
                },
                "validation": {"docstatus": ["=", 1]},
            },
            "Stock Entry Detail": {
                "doctype": "Stock Entry Detail",
                "field_map": {
                    "name": "ste_detail",
                    "parent": "against_stock_entry",
                    "serial_no": "serial_no",
                    "batch_no": "batch_no",
                },
                "postprocess": update_item,
                "condition": lambda doc: flt(doc.qty) - flt(doc.transferred_qty) > 0.00001,
            },
        },
        target_doc,
        set_missing_values,
    )
    doclist.to_warehouse = to_warehouse
    
    return doclist

File: maliki/public/js/stock_entry.js
let settings;
let parent_per_transferred_value;

frappe.ui.form.on('Stock Entry', {
    before_save(frm) {
        if(frm.doc.outgoing_stock_entry){
            if(frm.doc.to_warehouse === settings['damaged_warehouse']){
                frm.set_value('stock_title', 'الى مخزن الأضرار')
            }
            else{
                frm.set_value('stock_title', 'استلام في تركيا')
            }
        }
        else{
            frm.set_value('stock_title', 'الى النقل')
        }
    },
    async before_submit(frm) {
        if(!frm.doc.outgoing_stock_entry && frm.doc.add_to_transit){
            frm.set_value('transfer_status', "انتظار");
            return;
        }
        if(frm.doc.outgoing_stock_entry && (parent_per_transferred_value > 0 && parent_per_transferred_value < 100)){
            const status = await prompt_user();
            frm.set_value('issue_type', status);
            frm.set_value('transfer_status', 'استلام جزئي');
        }
    },
    async on_submit(frm) {
        if(frm.doc.outgoing_stock_entry){
            const parent_per_transferred = await frappe.db.get_value('Stock Entry', frm.doc.outgoing_stock_entry, 'per_transferred');
            const parent_per_transferred_value = parent_per_transferred.message.per_transferred;

            if(parent_per_transferred_value > 0 && parent_per_transferred_value < 100){
                parent_doc = await frappe.db.get_doc('Stock Entry', frm.doc.outgoing_stock_entry);
                frappe.confirm('يوجد أصناف غير مستلمة, هل تريد تحويلها لمخزن الأضرار؟',
                    () => {
                        // action to perform if Yes is selected
                        const parentfrm = frm;
                        parentfrm.doc = parent_doc;
                        parentfrm.docname = parent_doc.name;

                        frappe.model.open_mapped_doc({
                            method: "maliki.maliki.overrides.stock_entry.make_dmg_stock_in_entry",
                            frm: parentfrm,
                        });
                    }, () => {
                        // action to perform if No is selected
                    })
            }
        }
        if(['ضرر', 'ضياع'].includes(frm.doc.issue_type)){
            frm.call('set_custom_status')
        }
    },

    onload(frm) {
        if(!frm.doc.outgoing_stock_entry && frm.doc.__islocal && !frm.doc.stock_entry_type){
            frm.set_value('stock_entry_type', "Material Transfer");
        }
    },
    add_to_transit(frm) {
        if(frm.is_new() && settings){
            if(frm.doc.add_to_transit && !frm.doc.outgoing_stock_entry){
                frm.set_value('to_warehouse', settings["transit_warehouse"]);
            }else{
                frm.set_value('to_warehouse', settings["tr_warehouse"]);
            }
        }
    },
	async refresh(frm) {
        if(!frm.doc.additional_costs_je && frm.doc.docstatus === 1){
            frm.add_custom_button(__("تسجيل مصاريف الشحن"), function () {
                frm.call('register_additional_costs_je').then(() => {
                    frm.refresh();
                })
            });
        }

		settings = await frappe.db.get_doc('Maliki Settings', null);
        
        const per_transferred = frm.doc.per_transferred;
        const parent_per_transferred = await frappe.db.get_value('Stock Entry', frm.doc.outgoing_stock_entry, 'per_transferred');
        parent_per_transferred_value = parent_per_transferred.message.per_transferred;

        if(per_transferred > 0 && per_transferred < 100 && !frm.doc.__islocal){
            frm.add_custom_button(__("ارسال لمخزن الأضرار"), function () {
                
                frappe.model.open_mapped_doc({
                    method: "maliki.maliki.overrides.stock_entry.make_dmg_stock_in_entry",
                    frm: frm,
                });
            });
        }

        if(!frm.doc.__islocal){
            return;
        }
        
        if(parent_per_transferred_value > 0){
            return;
        }
        
        if(!frm.doc.outgoing_stock_entry){
            frm.set_value('add_to_transit', 1);
            frm.set_value('from_warehouse', settings["ly_warehouse"]);
        }else{
            frm.set_value('from_warehouse', settings["transit_warehouse"]);
            frm.set_value('to_warehouse', settings["tr_warehouse"]);
            if(frm.doc.additional_costs.length === 0){
                let new_row = frm.add_child('additional_costs');
                frappe.model.set_value(new_row.doctype, new_row.name, "expense_account", settings["shipping_account"]);
                frappe.model.set_value(new_row.doctype, new_row.name, "description", 'تكلفة النقل');
                frappe.model.set_value(new_row.doctype, new_row.name, "amount", 0);
                frm.refresh_field('additional_costs');
            }
            
        }
	}
})

function prompt_user() {
    return new Promise((resolve) => {
        frappe.prompt(
            [
                {
                    label: 'الحالة',
                    fieldname: 'custom_status',
                    fieldtype: 'Select',
                    options: ['ضرر', 'ضياع']
                }
            ],
            function(values) {
                resolve(values.custom_status);
            },
            'الرجاء ادخال نوع الاستلام',
            "حفظ"
        );
    });
}